src/assets/TokenDetectionController.ts

```diff
@@ -53,87 +57,157 @@ export class TokenDetectionController extends BaseController<
+      getPreferencesState();
+
     super(config, state);
     this.defaultConfig = {
       interval: DEFAULT_INTERVAL,
-      networkType: MAINNET,
       selectedAddress: '',
-      tokens: [],
+      disabled: true,
+      chainId: defaultChainId,
+      isDetectionEnabledFromPreferences: defaultUseTokenDetection,
+      isDetectionEnabledForNetwork:
+        isTokenDetectionSupportedForNetwork(defaultChainId),
+      ...config,
     };
+
     this.initialize();
     this.getTokensState = getTokensState;
     this.getTokenListState = getTokenListState;
-    this.addTokens = addTokens;
-    onTokensStateChange(({ tokens }) => {
-      this.configure({ tokens });
+    this.addDetectedTokens = addDetectedTokens;
+    this.getBalancesInSingleCall = getBalancesInSingleCall;
+
+    onTokenListStateChange(({ tokenList }) => {
+      const {
+        disabled,
+        isDetectionEnabledForNetwork,
+        isDetectionEnabledFromPreferences,
+      } = this.config;
+      const hasTokens = Object.keys(tokenList).length;
+
+      if (
+        !disabled &&
+        isDetectionEnabledForNetwork &&
+        isDetectionEnabledFromPreferences &&
+        hasTokens
+      ) {
+        this.detectTokens();
+      }
     });
 
-    onPreferencesStateChange(({ selectedAddress }) => {
-      const actualSelectedAddress = this.config.selectedAddress;
-      if (selectedAddress !== actualSelectedAddress) {
-        this.configure({ selectedAddress });
+    onPreferencesStateChange(({ selectedAddress, useTokenDetection }) => {
+      const {
+        disabled,
+        selectedAddress: currentSelectedAddress,
+        isDetectionEnabledFromPreferences,
+        isDetectionEnabledForNetwork,
+      } = this.config;
+      const isSelectedAddressChanged =
+        selectedAddress !== currentSelectedAddress;
+      const isDetectionChangedFromPreferences =
+        isDetectionEnabledFromPreferences !== useTokenDetection;
+
+      this.configure({
+        isDetectionEnabledFromPreferences: useTokenDetection,
+        selectedAddress,
+      });
+
+      if (
+        !disabled &&
+        isDetectionEnabledForNetwork &&
+        useTokenDetection &&
+        (isSelectedAddressChanged || isDetectionChangedFromPreferences)
+      ) {
         this.detectTokens();
       }
     });
 
-    onNetworkStateChange(({ provider }) => {
-      this.configure({ networkType: provider.type });
+    onNetworkStateChange(({ provider: { chainId } }) => {
+      const {
+        disabled,
+        chainId: currentChainId,
+        isDetectionEnabledFromPreferences,
+      } = this.config;
+      const isDetectionEnabledForNetwork =
+        isTokenDetectionSupportedForNetwork(chainId);
+      const isChainIdChanged = currentChainId !== chainId;
+
+      this.configure({
+        chainId,
+        isDetectionEnabledForNetwork,
+      });
+
+      if (
+        !disabled &&
+        isDetectionEnabledFromPreferences &&
+        isDetectionEnabledForNetwork &&
+        isChainIdChanged
```

Comment:
Need a test to cover `isChainIdChanged`. We could also remove some of these conditions, as they're redundant with the checks in `detectToken`